### PR TITLE
Add option for custom actions to Semantics widget

### LINKED ISSUE #12434: Semantics widget should allow specifying custom actions
Some developers want to specify custom a11y actions for their app that are different from the default a11y actions of a widget. For this, it would be cool if the Semantics widget would allow specifying a set of SemanticActions and handlers.

Example: One dev wants to expose a drop down menu as an "Adjustable" where swiping up and down switches between the options.

Related: https://github.com/flutter/flutter/issues/12433

### PR DESCRIPTION
**Pull request by goderbauer**
Fixes https://github.com/flutter/flutter/issues/12434.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 1]
```diff
@@ -3317,6 +3333,102 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
     markNeedsSemanticsUpdate(onlyLocalUpdates: (value != null) == hadValue);
   }
 
+  /// The handler for [SemanticsAction.tap].
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Can we have more details on each of these? What the semantics of the action are, examples of how the user might trigger it, and so on?

This is the main entry-point to this API so you cannot assume that the user will know about [SemanticsAction].

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 2]
```diff
@@ -3334,8 +3448,65 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
...
+    if (onScrollDown != null)
+      config.addAction(SemanticsAction.scrollDown, _performScrollDown);
+    if (onIncrease != null)
+      config.addAction(SemanticsAction.increase, _performIncrease);
+    if (onDecrease != null)
+      config.addAction(SemanticsAction.decrease, _performDecrease);
+  }
+
+  void _performTap() {
+    if (onTap != null)
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: i guess the concern is that the system will invoke our callbacks sometime after we've been told we no longer have an onTap handler?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -4540,6 +4548,30 @@ class Semantics extends SingleChildRenderObjectWidget {
     return textDirection ?? (label != null || value != null || hint != null ? Directionality.of(context) : null);
   }
 
+  /// The handler for [SemanticsAction.tap].
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: copy all the documentation you add for the callbacks above into these too, they're even more the front-line docs.

File: packages/flutter/test/widgets/semantics_test.dart [hunk 4]
```diff
@@ -370,4 +370,147 @@ void main() {
...
+
+    semantics.dispose();
+  });
+
+  testWidgets('Actions can be replaced without triggering semantics update', (WidgetTester tester) async {
+    final SemanticsTester semantics = new SemanticsTester(tester);
+    int semanticsUpdateCount = 0;
+    tester.binding.pipelineOwner.ensureSemantics(
+      listener: () {
+        semanticsUpdateCount++;
```
Comment (Hixie) [inline packages/flutter/test/widgets/semantics_test.dart]: `foo += 1` or `++foo`

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: Done.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/basic.dart]: Done.

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: That was my concern. Does the system guarantee that that will never happen?

Comment (Hixie) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: At the render level, I guess not. You can modify a render object, then drive the pipeline much later.

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 5]
```diff
@@ -3317,6 +3333,162 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
...
+
+  /// The handler for [SemanticsAction.scrollUp].
+  ///
+  /// This is the semantic equivalent of a user moving their finger across the
+  /// screen from bottom to top. It should be recognized by controls that are
+  /// vertically scrollable.
+  ///
+  /// VoiceOver users on iOS can trigger this action by swiping up with three
+  /// fingers. TalkBack users on Android can trigger this action by swiping
+  /// right and then left in one motion path.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: right then left to go up? isn't that left?

Comment (Hixie) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: if this is correct, this should talk about disambiguation with onScrollLeft

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 6]
```diff
@@ -3317,6 +3333,162 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
...
+
+  /// The handler for [SemanticsAction.scrollDown].
+  ///
+  /// This is the semantic equivalent of a user moving their finger across the
+  /// screen from top to bottom. It should be recognized by controls that are
+  /// vertically scrollable.
+  ///
+  /// VoiceOver users on iOS can trigger this action by swiping down with three
+  /// fingers. TalkBack users on Android can trigger this action by swiping
+  /// left and then right in one motion path.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: ditto

**Comment by Hixie** [comment]
LGTM modulo minor doc nits (don't forget to propagate those to the widgets too)

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: In the API level we are using android only support two scroll directions: forward and backwards. Both up and and left map to forward. Similarly, right and down map to backwards.

SDK v23+ differentiates between the four.

**Pull request merged**